src/framework/audio/internal/plugins/registeraudiopluginsscenario.cpp

```diff
@@ -0,0 +1,145 @@
+/*
+ * SPDX-License-Identifier: GPL-3.0-only
+ * MuseScore-CLA-applies
+ *
+ * MuseScore
+ * Music Composition & Notation
+ *
+ * Copyright (C) 2021 MuseScore BVBA and others
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include "registeraudiopluginsscenario.h"
+
+#include <QApplication>
+
+#include "translation.h"
+#include "log.h"
+
+using namespace mu::audio;
+using namespace mu::framework;
+
+void RegisterAudioPluginsScenario::init()
+{
+    m_progress.finished.onReceive(this, [this](const ProgressResult& res) {
+        if (res.ret.code() == static_cast<int>(Ret::Code::Cancel)) {
+            m_aborted = true;
+        }
+    });
+}
+
+mu::Ret RegisterAudioPluginsScenario::registerNewPlugins()
+{
+    TRACEFUNC;
+
+    io::paths_t newPluginPaths;
+
+    for (IAudioPluginsScannerPtr scanner : scannerRegister()->scanners()) {
+        io::paths_t paths = scanner->scanPlugins();
+
+        for (const io::path_t& path : paths) {
+            if (!knownPluginsRegister()->exists(path)) {
+                newPluginPaths.push_back(path);
+            }
+        }
+    }
+
+    startPluginsRegistration(newPluginPaths);
+
+    return make_ok();
+}
+
+void RegisterAudioPluginsScenario::startPluginsRegistration(const io::paths_t& pluginPaths)
+{
+    if (pluginPaths.empty()) {
+        return;
+    }
+
+    Ret ret = interactive()->showProgress(mu::trc("audio", "Scanning audio plugins"), &m_progress);
+    if (!ret) {
+        LOGE() << ret.toString();
+    }
+
+    m_aborted = false;
+    m_progress.started.notify();
+
+    std::string appPath = globalConfiguration()->appBinPath().toStdString();
+    int64_t pluginCount = static_cast<int64_t>(pluginPaths.size());
+
+    for (int64_t i = 0; i < pluginCount; ++i) {
+        if (m_aborted) {
+            return;
+        }
+
+        const io::path_t& pluginPath = pluginPaths[i];
+        std::string pluginPathStr = pluginPath.toStdString();
+
+        m_progress.progressChanged.send(i, pluginCount, pluginPathStr);
+        qApp->processEvents();
+
+        int code = process()->execute(appPath, { "--audio-plugin-probe", pluginPathStr });
+        if (code == 0) {
+            continue;
+        }
+
+        AudioPluginInfo info;
+        info.meta.id = io::filename(pluginPath).toStdString();
+        info.path = pluginPath;
+        info.enabled = false;
+        info.errorCode = code;
+
+        Ret ret = knownPluginsRegister()->registerPlugin(info);
```

Comment:
It seems to me that "successful" plugins should also be written here. Otherwise, it's not "those we know about", but something like "failure plugins"

Reply:
looked more...
as I understand it, "successful" are register in a separate process (probe), and "not successful" are register in the main process - I believe that this can lead to misunderstandings.
I think that both should be registered in one place.
Apparently, it is better to do this in a separate process, i.e. there should be something like this:

// main proccess
```
...
const io::path_t& pluginPath = pluginPaths.at(i);
...
process()->execute(appPath, { "--audio-plugin-registre", pluginPathStr });
```

// separate process
```
    case framework::IApplication::RunMode::AudioPluginRegistre: {
        io::path_t pluginPath = commandLine.audioPluginPath();

        QMetaObject::invokeMethod(qApp, [pluginPath]() {
                LOGD() << "Scanning audio plugin: " << pluginPath;
                qApp->exit(0);
        QMetaObject::invokeMethod(qApp, [this, pluginPath]() {
                Ret ret = registerAudioPluginsScenario()->registerPlugin(pluginPath);
                if (!ret) {
                    LOGE() << ret.toString();
                }
                qApp->exit(ret.code());
            }, Qt::QueuedConnection);


mu::Ret RegisterAudioPluginsScenario::registerPlugin(const io::path_t& pluginPath)
{
    AudioPluginInfo info;
    RetVal<AudioResourceMeta> meta = reader->readMeta(pluginPath);
    if (meta.ret) {
       make success info...
    } else {
        LOGE() << meta.ret.toString();
        make fail info...
    } 

    Ret ret = knownPluginsRegister()->registerPlugin(info);
    return ret;
}
```